packages/sdk/src/modules/vault/Vault.ts

```diff
@@ -244,40 +245,65 @@ export class Vault extends Predicate<[]> {
       () => FAKE_WITNESSES,
     );
     transactionRequest.witnesses.push(...fakeSignatures);
-    const transactionCost = await this.getTransactionCost(transactionRequest);
-    transactionRequest.maxFee = transactionCost.maxFee;
-    transactionRequest = await this.fund(transactionRequest, transactionCost);
-
-    let totalGasUsed = bn(0);
-    transactionRequest.inputs.forEach((input) => {
-      if ('predicate' in input && input.predicate) {
-        input.witnessIndex = 0;
-        input.predicateGasUsed = undefined;
-        totalGasUsed = totalGasUsed.add(predicateGasUsed);
+
+    const fundAndEstimateMaxFee = async () => {
+      const transactionCost = await this.getTransactionCost(transactionRequest);
+      if (transactionRequest.maxFee.lt(transactionCost.maxFee)) {
+        transactionRequest.maxFee = transactionCost.maxFee;
       }
-    });
+      if ('gasLimit' in transactionRequest) {
+        transactionRequest.gasLimit = transactionCost.gasUsed;
+      }
```

Comment:
Here you should also check, if `transactionRequest.gasLimit.le(transactionCost.gasUsed)` in this way if the application provides a bigger gasLimit you should assign it as the gasLimit.